### PR TITLE
Add support for values params for functions

### PR DESCRIPTION
**Pull request by dluc**
### Motivation and Context

1. When writing a prompt one might want to use the symbols `{{` and `}}` without triggering the template engine.
2. When invoking functions, one might want to pass a value rather than a variable, e.g. `{{ recall "my preferred browser" }}`.
3. Semantic functions might want to call other semantic functions.

### Description

This PR adds support for "values", so special syntax can be embedded, and so that functions can be invoked with predefined values without the need to create a temporary variable.

E.g.

```
{{ "{{" }} and {{ "}}" }} are special SK sequences.
```
```
... {{ read 'c:\documents\ai\file.txt' }} ...
```
```
... {{ "quotes' \"escaping\" example" }} ...
```
```
... {{ 'quotes\' "escaping" example' }} ...
```
```
{{ 'two special chars \\\' here' }}
```

Added missing tests for all the template language, including end to end integration tests covering several edge cases.

Also: semantic functions can now call other semantic functions. The previous implementation would allow semantic functions to call only native functions.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [x] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/SemanticKernel/SkillDefinition/IReadOnlySkillCollection.cs [hunk 1]
```diff
@@ -49,6 +49,28 @@ public interface IReadOnlySkillCollection
     /// <returns>True if the function exists</returns>
     bool HasNativeFunction(string functionName);
 
+    /// <summary>
+    /// Return the function delegate stored in the collection, regardless of the function type
+    /// </summary>
+    /// <param name="functionName">Function name</param>
+    /// <returns>Semantic function delegate</returns>
```
Comment (lemillermicrosoft) [inline dotnet/src/SemanticKernel/SkillDefinition/IReadOnlySkillCollection.cs:56]: Could be native, right? Same below.

File: dotnet/src/SemanticKernel/TemplateEngine/PromptTemplateEngine.cs [hunk 2]
```diff
@@ -24,86 +23,77 @@ namespace Microsoft.SemanticKernel.TemplateEngine;
...
-    public async Task<string> RenderAsync(
-        string templateText,
-        SKContext executionContext)
+    /// <inheritdoc/>
+    public async Task<string> RenderAsync(string templateText, SKContext context)
     {
         this._log.LogTrace("Rendering string template: {0}", templateText);
         var blocks = this.ExtractBlocks(templateText);
-        return await this.RenderAsync(blocks, executionContext);
+        return await this.RenderAsync(blocks, context);
```
Comment (lemillermicrosoft) [inline dotnet/src/SemanticKernel/TemplateEngine/PromptTemplateEngine.cs:61]: May be out of scope of this review, but I'm seeing first hand and getting reports from folks that SK behaves differently than the playground with identical prompts and it seems to come down to '\r\n' vs '\n' - I'm wondering if we should go ahead and make that part of the engine to convert everything for consistency?

File: dotnet/src/SemanticKernel/TemplateEngine/Blocks/VarBlock.cs [hunk 3]
```diff
@@ -57,32 +68,16 @@ internal override string Render(ContextVariables? variables)
...
-    {
-        return Regex.IsMatch(text, "^[a-zA-Z0-9_]*$");
-    }
+        var exists = variables.Get(this.Name, out string value);
+        if (!exists) { this.Log.LogWarning("Variable `{0}{1}` not found", Symbols.VarPrefix, this.Name); }
 
-    private string VarName()
-    {
-        return this.Content.Length < 2 ? "" : this.Content[1..];
+        return exists ? value : "";
```
Comment (shawncal) [inline dotnet/src/SemanticKernel/TemplateEngine/Blocks/VarBlock.cs:81]: nit string.Empty

File: dotnet/src/SemanticKernel/TemplateEngine/IPromptTemplateEngine.cs [hunk 4]
```diff
@@ -27,21 +27,21 @@ public interface IPromptTemplateEngine
...
-    /// <param name="executionContext">Access into the current kernel execution context</param>
+    /// <param name="context">Access into the current kernel execution context</param>
     /// <returns>The prompt template ready to be used for an AI request</returns>
     Task<string> RenderAsync(
         string templateText,
-        SKContext executionContext);
+        SKContext context);
 
     /// <summary>
     /// Given a a list of blocks render each block and compose the final result
```
Comment (shawncal) [inline dotnet/src/SemanticKernel/TemplateEngine/IPromptTemplateEngine.cs:37]: Can you fix this while you're in the code pls? ("Given a a...")

File: docs/PROMPT_TEMPLATE_LANGUAGE.md [hunk 5]
```diff
@@ -132,3 +142,112 @@ async Task<string> GenResponseToEmailAsync(
...
+Both are rendered to:
+
+    ... quotes' "escaping" example ...
+
+Note that for consistency, the sequences **«`\'`»** and **«`\"`»** do always render
+to **«`'`»** and **«`"`»**, even when escaping might not be required.
+
+For instance:
+
+    ... {{ 'no need to \"escape" ' }} ...
```
[the remark this answers is not in the document]
Comment (dluc) [reply docs/PROMPT_TEMPLATE_LANGUAGE.md:206]: it's on purpose, see below line 208..210

File: dotnet/src/IntegrationTest/IntegrationTests.csproj [hunk 6]
```diff
@@ -39,4 +39,11 @@
     </None>
   </ItemGroup>
 
+  <ItemGroup>
+    <None Remove="TemplateLanguage\tests.txt" />
+    <Content Include="TemplateLanguage\tests.txt">
```
[the remark this answers is not in the document]
Comment (dluc) [reply dotnet/src/IntegrationTest/IntegrationTests.csproj:44]: probly not, checking

File: dotnet/src/SemanticKernel/TemplateEngine/Blocks/CodeBlock.cs [hunk 7]
```diff
@@ -13,86 +10,66 @@
 
 namespace Microsoft.SemanticKernel.TemplateEngine.Blocks;
 
+#pragma warning disable CA2254 // error strings are used also internally, not just for logging
```
[the remark this answers is not in the document]
Comment (dluc) [reply dotnet/src/SemanticKernel/TemplateEngine/Blocks/CodeBlock.cs:13]: yeah it's on purpose, aside from test classes, I try to reduce the scope of these exception to a minimum

File: dotnet/src/SemanticKernel/TemplateEngine/Blocks/FunctionIdBlock.cs [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    internal string FunctionName { get; } = string.Empty;
+
+    internal FunctionIdBlock(string? text, ILogger? log = null)
+        : base(text?.Trim(), log)
+    {
+        var functionNameParts = this.Content.Split('.');
+        if (functionNameParts.Length > 2)
+        {
+            this.Log.LogError("Invalid function name `{0}`", this.Content);
+            return;
```
[the remark this answers is not in the document]
Comment (dluc) [reply dotnet/src/SemanticKernel/TemplateEngine/Blocks/FunctionIdBlock.cs:27]: fixing - the code was throwing in IsValid, better doing it here

File: dotnet/src/IntegrationTest/IntegrationTests.csproj (hunk 6 above)
[the remark this answers is not in the document]
Comment (dluc) [reply dotnet/src/IntegrationTest/IntegrationTests.csproj:40]: removed

File: dotnet/src/SemanticKernel/TemplateEngine/Blocks/Block.cs [hunk 9]
```diff
@@ -25,18 +30,24 @@ public abstract class Block
...
     {
-        throw new NotImplementedException("This block doesn't support code execution");
+        throw new NotImplementedException("This block doesn't support async code execution, use Render(ContextVariables) instead.");
     }
 
-    internal abstract bool IsValid(out string error);
+    internal virtual string Render(ContextVariables? variables)
+    {
+        throw new NotImplementedException("This block requires async code execution, use RenderCodeAsync(SKContext) instead.");
+    }
```
[the remark this answers is not in the document]
Comment (dluc) [reply dotnet/src/SemanticKernel/TemplateEngine/Blocks/Block.cs:50]: nice idea, thanks

[the remark this answers is not in the document]
Comment (dluc) [reply dotnet/src/SemanticKernel/TemplateEngine/Blocks/Block.cs]: done

**Review by lemillermicrosoft — APPROVED** [review]
Did not have chance to test yet, but LGTM.

File: dotnet/src/SemanticKernel/TemplateEngine/Blocks/Block.cs [hunk 10]
```diff
@@ -25,18 +27,14 @@ public abstract class Block
...
-    internal virtual Task<string> RenderCodeAsync(SKContext executionContext)
-    {
-        throw new NotImplementedException("This block doesn't support code execution");
+        this.Content = content ?? string.Empty;
     }
 
-    internal abstract bool IsValid(out string error);
-
-    internal abstract string Render(ContextVariables? variables);
+    public abstract bool IsValid(out string error);
```
Comment (shawncal) [inline dotnet/src/SemanticKernel/TemplateEngine/Blocks/Block.cs:39]: Got some new warnings here

**Review by shawncal — APPROVED** [review]

**Comment by shawncal** [comment]
@dluc Merge conflict (renamed tests folder) and a new warning. Happy to merge when ready

**Review by lemillermicrosoft — COMMENTED** [review]
Had one question about variable expansion on function name support.

File: docs/PROMPT_TEMPLATE_LANGUAGE.md [hunk 11]
```diff
@@ -57,26 +61,31 @@ For instance, the code above is equivalent to:
 ## Function parameters
 
 To call an external function and pass a parameter to it, use the
-`{namespace.functionName $varName}` syntax.
+`{{namespace.functionName $varName}}` and
+`{{namespace.functionName "value"}}` syntax.
```
[the remark this answers is not in the document]
Comment (dluc) [reply docs/PROMPT_TEMPLATE_LANGUAGE.md:65]: correct, it's not supported. Variables can be used only to pass values, not to generate names/code. For more advanced coding, we rely on native functions

**Comment by dluc** [comment]
@shawncal @lemillermicrosoft warnings fixed / comments addressed

**Review by shawncal — APPROVED** [review]
